File: src/priox/md/bridge/water.py
"""Water model definitions and parameterization logic (Explicit Solvent)."""

from __future__ import annotations

import dataclasses
from dataclasses import dataclass
from typing import Dict, List, Tuple, Optional
import math

# Conversion factors
DEG2RAD = math.pi / 180.0

@dataclass
class WaterModelParams:
    """Parameters for a specific water model."""
    name: str
    atoms: List[str]
    charges: Dict[str, float]
    sigmas: Dict[str, float]  # Angstroms
    epsilons: Dict[str, float] # kcal/mol
    bonds: List[Tuple[str, str, float, float]] # (atom1, atom2, length, k_kcal_mol_A2)
    angles: List[Tuple[str, str, str, float, float]] # (a1, a2, a3, theta_rad, k_kcal_mol_rad2)
    # Constraints for rigid water (SETTLE/RATTLE)
    # Usually we constrain OH bonds and HH distance (or HOH angle)
    # Stored as (atom1, atom2, length)
    constraints: List[Tuple[str, str, float]] 
    
    # Virtual Sites (for 4/5-site models)
    # List of dicts matching structure in ForceField.virtual_sites
    # {"type": "Twocenter", "siteName": "M", "atoms": ["O", "H1", "H2"], ... parameters ...}
    virtual_sites: Optional[List[dict]] = None

# -------------------------------------------------------------------------
# TIP3P (Standard)
# -------------------------------------------------------------------------
# r(OH) = 0.9572 A, theta(HOH) = 104.52 deg
# HH dist = 2 * 0.9572 * sin(104.52/2) = 1.51390065
TIP3P_DEFS = WaterModelParams(
    name="TIP3P",
    atoms=["O", "H1", "H2"],
    charges={"O": -0.834, "H1": 0.417, "H2": 0.417},
    sigmas={"O": 3.15061, "H1": 0.0001, "H2": 0.0001},
    epsilons={"O": 0.1521, "H1": 0.0, "H2": 0.0},
    bonds=[
        ("O", "H1", 0.9572, 450.0),
        ("O", "H2", 0.9572, 450.0)
    ],
    angles=[
        ("H1", "O", "H2", 104.52 * DEG2RAD, 100.0)
    ],
    constraints=[
        ("O", "H1", 0.9572),
        ("O", "H2", 0.9572),
        ("H1", "H2", 1.51390065)
    ],
    virtual_sites=None
)

# -------------------------------------------------------------------------
# SPC/E (Extended Simple Point Charge)
# -------------------------------------------------------------------------
# r(OH) = 1.0 A, theta(HOH) = 109.47 deg
# HH dist = 2 * 1.0 * sin(109.47/2) = 1.63298
SPCE_DEFS = WaterModelParams(
    name="SPCE",
    atoms=["O", "H1", "H2"],
    charges={"O": -0.8476, "H1": 0.4238, "H2": 0.4238},
    sigmas={"O": 3.166, "H1": 0.0001, "H2": 0.0001},
    epsilons={"O": 0.1553, "H1": 0.0, "H2": 0.0},
    bonds=[
        ("O", "H1", 1.0, 450.0),
        ("O", "H2", 1.0, 450.0)
    ],
    angles=[
        ("H1", "O", "H2", 109.47 * DEG2RAD, 100.0)
    ],
    constraints=[
        ("O", "H1", 1.0),
        ("O", "H2", 1.0),
        ("H1", "H2", 1.63298086)
    ],
    virtual_sites=None
)

# -------------------------------------------------------------------------
# TIP4P-Ew (Ewald optimized)
# -------------------------------------------------------------------------
# 4-site model with virtual site M
# r(OH) = 0.9572, theta = 104.52
# M is on bisector of HOH, distance d_OM = 0.125 A from O towards H's?
# Charges: H = 0.52422, M = -1.04844, O = 0.0
# Warning: O has charge 0, but M carries the negative charge.
# Sigma/Eps on O.
TIP4PEW_DEFS = WaterModelParams(
    name="TIP4PEW",
    atoms=["O", "H1", "H2", "M"], # M is virtual
    charges={"O": 0.0, "H1": 0.52422, "H2": 0.52422, "M": -1.04844},
    sigmas={"O": 3.16435, "H1": 0.0001, "H2": 0.0001, "M": 0.0001},
    epsilons={"O": 0.16275, "H1": 0.0, "H2": 0.0, "M": 0.0},
    bonds=[
        ("O", "H1", 0.9572, 450.0),
        ("O", "H2", 0.9572, 450.0)
    ],
    angles=[
        ("H1", "O", "H2", 104.52 * DEG2RAD, 100.0)
    ],
    constraints=[
        ("O", "H1", 0.9572),
        ("O", "H2", 0.9572),
        ("H1", "H2", 1.51390065)
    ],
    virtual_sites=[
        # Type depends on implementation.
        # "ThreeAtomAverage" / "LocalCoordinates" / "TiP4P" specific
        # Prolix system.py has: 'siteName', 'atoms', parameters p, wo, wx, wy...
        # We need to map to what system.py expects from FF.
        # For now, placeholder. TIP4P usually handled by specialized VS logic.
    ]
)


def get_water_params(model_name: str = "TIP3P", rigid: bool = True) -> WaterModelParams:
    """Get parameters for specified water model name (case-insensitive).
    
    Args:
        model_name: Water model name (TIP3P, SPCE, TIP4PEW, etc.)
        rigid: If True, sets bond and angle force constants to 0.0, matching
               OpenMM's rigidWater=True behavior. Constraints still define
               geometry for SHAKE/SETTLE. Default True for standard MD.
    
    Returns:
        WaterModelParams with appropriate force constants.
    """
    name = model_name.upper()
    if name == "TIP3P":
        base = TIP3P_DEFS
    elif name == "SPCE" or name == "SPC/E":
        base = SPCE_DEFS
    elif name == "TIP4PEW" or name == "TIP4P-EW":
        base = TIP4PEW_DEFS
    elif name == "HOH" or name == "WAT" or name == "SOL":
        # Default fallback
        base = TIP3P_DEFS
    else:
        raise ValueError(f"Unknown water model: {model_name}")
    
    if rigid:
        # Zero out bond and angle force constants
        # Keep equilibrium values for constraint geometry
        rigid_bonds = [(a1, a2, length, 0.0) for a1, a2, length, _ in base.bonds]
        rigid_angles = [(a1, a2, a3, theta, 0.0) for a1, a2, a3, theta, _ in base.angles]
        
        return dataclasses.replace(
            base,
            bonds=rigid_bonds,
            angles=rigid_angles
        )
    
    return base

File: crates/proxide-physics/src/physics/water.rs
//! Water model definitions and parameterization.
//!
//! Provides standard water model parameters (TIP3P, SPC/E, TIP4P-Ew)
//! for explicit solvent simulations.

use std::collections::HashMap;

/// Parameters for a specific water model.
#[derive(Debug, Clone)]
pub struct WaterModel {
    pub name: String,
    pub atoms: Vec<String>,
    pub charges: HashMap<String, f32>,
    pub sigmas: HashMap<String, f32>,   // Angstroms
    pub epsilons: HashMap<String, f32>, // kcal/mol
    /// Bonds: (atom1_name, atom2_name, equilibrium_length, force_constant)
    pub bonds: Vec<(String, String, f32, f32)>,
    /// Angles: (a1, a2, a3, theta_radians, force_constant)
    pub angles: Vec<(String, String, String, f32, f32)>,
    /// Constraints: (atom1_name, atom2_name, distance)
    pub constraints: Vec<(String, String, f32)>,
    /// Whether this model has virtual sites (e.g., TIP4P)
    pub has_virtual_sites: bool,
}

// Conversion factor
const DEG2RAD: f32 = std::f32::consts::PI / 180.0;

/// Get parameters for TIP3P water model.
fn tip3p() -> WaterModel {
    let mut charges = HashMap::new();
    charges.insert("O".to_string(), -0.834);
    charges.insert("H1".to_string(), 0.417);
    charges.insert("H2".to_string(), 0.417);

    let mut sigmas = HashMap::new();
    sigmas.insert("O".to_string(), 3.15061);
    sigmas.insert("H1".to_string(), 0.0001);
    sigmas.insert("H2".to_string(), 0.0001);

    let mut epsilons = HashMap::new();
    epsilons.insert("O".to_string(), 0.1521);
    epsilons.insert("H1".to_string(), 0.0);
    epsilons.insert("H2".to_string(), 0.0);

    // r(OH) = 0.9572 Å, theta(HOH) = 104.52°
    // HH dist = 2 * 0.9572 * sin(104.52/2) ≈ 1.5139
    WaterModel {
        name: "TIP3P".to_string(),
        atoms: vec!["O".to_string(), "H1".to_string(), "H2".to_string()],
        charges,
        sigmas,
        epsilons,
        bonds: vec![
            ("O".to_string(), "H1".to_string(), 0.9572, 450.0),
            ("O".to_string(), "H2".to_string(), 0.9572, 450.0),
        ],
        angles: vec![(
            "H1".to_string(),
            "O".to_string(),
            "H2".to_string(),
            104.52 * DEG2RAD,
            100.0,
        )],
        constraints: vec![
            ("O".to_string(), "H1".to_string(), 0.9572),
            ("O".to_string(), "H2".to_string(), 0.9572),
            ("H1".to_string(), "H2".to_string(), 1.513_900_6),
        ],
        has_virtual_sites: false,
    }
}

/// Get parameters for SPC/E water model.
fn spce() -> WaterModel {
    let mut charges = HashMap::new();
    charges.insert("O".to_string(), -0.8476);
    charges.insert("H1".to_string(), 0.4238);
    charges.insert("H2".to_string(), 0.4238);

    let mut sigmas = HashMap::new();
    sigmas.insert("O".to_string(), 3.166);
    sigmas.insert("H1".to_string(), 0.0001);
    sigmas.insert("H2".to_string(), 0.0001);

    let mut epsilons = HashMap::new();
    epsilons.insert("O".to_string(), 0.1553);
    epsilons.insert("H1".to_string(), 0.0);
    epsilons.insert("H2".to_string(), 0.0);

    // r(OH) = 1.0 Å, theta(HOH) = 109.47°
    // HH dist = 2 * 1.0 * sin(109.47/2) ≈ 1.6330
    WaterModel {
        name: "SPCE".to_string(),
        atoms: vec!["O".to_string(), "H1".to_string(), "H2".to_string()],
        charges,
        sigmas,
        epsilons,
        bonds: vec![
            ("O".to_string(), "H1".to_string(), 1.0, 450.0),
            ("O".to_string(), "H2".to_string(), 1.0, 450.0),
        ],
        angles: vec![(
            "H1".to_string(),
            "O".to_string(),
            "H2".to_string(),
            109.47 * DEG2RAD,
            100.0,
        )],
        constraints: vec![
            ("O".to_string(), "H1".to_string(), 1.0),
            ("O".to_string(), "H2".to_string(), 1.0),
            ("H1".to_string(), "H2".to_string(), 1.632_980_8),
        ],
        has_virtual_sites: false,
    }
}

/// Get parameters for TIP4P-Ew water model.
fn tip4pew() -> WaterModel {
    let mut charges = HashMap::new();
    charges.insert("O".to_string(), 0.0);
    charges.insert("H1".to_string(), 0.52422);
    charges.insert("H2".to_string(), 0.52422);
    charges.insert("M".to_string(), -1.04844);

    let mut sigmas = HashMap::new();
    sigmas.insert("O".to_string(), 3.16435);
    sigmas.insert("H1".to_string(), 0.0001);
    sigmas.insert("H2".to_string(), 0.0001);
    sigmas.insert("M".to_string(), 0.0001);

    let mut epsilons = HashMap::new();
    epsilons.insert("O".to_string(), 0.16275);
    epsilons.insert("H1".to_string(), 0.0);
    epsilons.insert("H2".to_string(), 0.0);
    epsilons.insert("M".to_string(), 0.0);

    WaterModel {
        name: "TIP4PEW".to_string(),
        atoms: vec![
            "O".to_string(),
            "H1".to_string(),
            "H2".to_string(),
            "M".to_string(),
        ],
        charges,
        sigmas,
        epsilons,
        bonds: vec![
            ("O".to_string(), "H1".to_string(), 0.9572, 450.0),
            ("O".to_string(), "H2".to_string(), 0.9572, 450.0),
        ],
        angles: vec![(
            "H1".to_string(),
            "O".to_string(),
            "H2".to_string(),
            104.52 * DEG2RAD,
            100.0,
        )],
        constraints: vec![
            ("O".to_string(), "H1".to_string(), 0.9572),
            ("O".to_string(), "H2".to_string(), 0.9572),
            ("H1".to_string(), "H2".to_string(), 1.513_900_6),
        ],
        has_virtual_sites: true,
    }
}

/// Get water model parameters by name.
///
/// Supported models: TIP3P, SPCE (SPC/E), TIP4PEW (TIP4P-Ew)
///
/// If `rigid` is true, bond and angle force constants are set to 0.0,
/// matching OpenMM's rigidWater=True behavior.
pub fn get_water_model(name: &str, rigid: bool) -> Result<WaterModel, String> {
    let upper = name.to_uppercase();
    let mut model = match upper.as_str() {
        "TIP3P" => tip3p(),
        "SPCE" | "SPC/E" => spce(),
        "TIP4PEW" | "TIP4P-EW" => tip4pew(),
        "HOH" | "WAT" | "SOL" => tip3p(), // Default fallback
        _ => return Err(format!("Unknown water model: {}", name)),
    };

    if rigid {
        // Zero out force constants, keep equilibrium values for constraints
        model.bonds = model
            .bonds
            .into_iter()
            .map(|(a1, a2, length, _)| (a1, a2, length, 0.0))
            .collect();
        model.angles = model
            .angles
            .into_iter()
            .map(|(a1, a2, a3, theta, _)| (a1, a2, a3, theta, 0.0))
            .collect();
    }

    Ok(model)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tip3p() {
        let model = get_water_model("TIP3P", false).unwrap();
        assert_eq!(model.name, "TIP3P");
        assert_eq!(model.atoms.len(), 3);
        assert!((model.charges["O"] - (-0.834)).abs() < 0.001);
        assert!(!model.has_virtual_sites);
    }

    #[test]
    fn test_tip4pew() {
        let model = get_water_model("TIP4PEW", false).unwrap();
        assert_eq!(model.atoms.len(), 4);
        assert!(model.has_virtual_sites);
        assert!((model.charges["M"] - (-1.04844)).abs() < 0.001);
    }

    #[test]
    fn test_rigid_water() {
        let model = get_water_model("TIP3P", true).unwrap();
        // Force constants should be 0
        for (_, _, _, k) in &model.bonds {
            assert_eq!(*k, 0.0);
        }
        for (_, _, _, _, k) in &model.angles {
            assert_eq!(*k, 0.0);
        }
        // But equilibrium values preserved
        assert!((model.bonds[0].2 - 0.9572).abs() < 0.001);
    }
}
